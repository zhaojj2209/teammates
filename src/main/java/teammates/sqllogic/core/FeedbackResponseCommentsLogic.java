package teammates.sqllogic.core;

import java.util.UUID;

import teammates.common.exception.EntityAlreadyExistsException;
import teammates.common.exception.EntityDoesNotExistException;
import teammates.common.exception.InvalidParametersException;
import teammates.storage.sqlapi.FeedbackResponseCommentsDb;
import teammates.storage.sqlentity.FeedbackResponseComment;
import teammates.ui.request.FeedbackResponseCommentUpdateRequest;

/**
 * Handles operations related to feedback response comments.
 *
 * @see FeedbackResponseComment
 * @see FeedbackResponseCommentsDb
 */
public final class FeedbackResponseCommentsLogic {

    private static final FeedbackResponseCommentsLogic instance = new FeedbackResponseCommentsLogic();
    private FeedbackResponseCommentsDb frcDb;

    private FeedbackResponseCommentsLogic() {
        // prevent initialization
    }

    public static FeedbackResponseCommentsLogic inst() {
        return instance;
    }

    /**
     * Initialize dependencies for {@code FeedbackResponseCommentsLogic}.
     */
    void initLogicDependencies(FeedbackResponseCommentsDb frcDb) {
        this.frcDb = frcDb;
    }

    /**
     * Gets an feedback response comment by feedback response comment id.
     * @param id of feedback response comment.
     * @return the specified feedback response comment.
     */
    public FeedbackResponseComment getFeedbackResponseComment(Long id) {
        return frcDb.getFeedbackResponseComment(id);
    }

    /**
     * Gets the comment associated with the response.
     */
    public FeedbackResponseComment getFeedbackResponseCommentForResponseFromParticipant(
            UUID feedbackResponseId) {
        return frcDb.getFeedbackResponseCommentForResponseFromParticipant(feedbackResponseId);
    }

    /**
     * Creates a feedback response comment.
     * @throws EntityAlreadyExistsException if the comment alreadty exists
     * @throws InvalidParametersException if the comment is invalid
     */
    public FeedbackResponseComment createFeedbackResponseComment(FeedbackResponseComment frc)
            throws InvalidParametersException, EntityAlreadyExistsException {
        return frcDb.createFeedbackResponseComment(frc);
    }

    /**
     * Deletes a feedbackResponseComment.
     */
    public void deleteFeedbackResponseComment(Long frcId) {
        frcDb.deleteFeedbackResponseComment(frcId);
    }

    /**
     * Updates a feedback response comment.
     * @throws EntityDoesNotExistException if the comment does not exist
     */
    public FeedbackResponseComment updateFeedbackResponseComment(Long frcId,
            FeedbackResponseCommentUpdateRequest updateRequest, String updaterEmail)
            throws EntityDoesNotExistException {
        FeedbackResponseComment comment = frcDb.getFeedbackResponseComment(frcId);
        if (comment == null) {
            throw new EntityDoesNotExistException("Trying to update a feedback response comment that does not exist.");
        }

        comment.setCommentText(updateRequest.getCommentText());
        comment.setShowCommentTo(updateRequest.getShowCommentTo());
        comment.setShowGiverNameTo(updateRequest.getShowGiverNameTo());
        comment.setLastEditorEmail(updaterEmail);

        return comment;
    }

    /**
     * Updates all email fields of feedback response comments.
     */
    public void updateFeedbackResponseCommentsEmails(String courseId, String oldEmail, String updatedEmail) {
        frcDb.updateGiverEmailOfFeedbackResponseComments(courseId, oldEmail, updatedEmail);
        frcDb.updateLastEditorEmailOfFeedbackResponseComments(courseId, oldEmail, updatedEmail);
    }

}
