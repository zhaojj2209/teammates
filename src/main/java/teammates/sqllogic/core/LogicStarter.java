package teammates.sqllogic.core;

import javax.servlet.ServletContextEvent;
import javax.servlet.ServletContextListener;

import teammates.common.util.Logger;
import teammates.storage.sqlapi.AccountRequestsDb;
import teammates.storage.sqlapi.AccountsDb;
import teammates.storage.sqlapi.CoursesDb;
import teammates.storage.sqlapi.DeadlineExtensionsDb;
import teammates.storage.sqlapi.FeedbackQuestionsDb;
import teammates.storage.sqlapi.FeedbackResponseCommentsDb;
import teammates.storage.sqlapi.FeedbackResponsesDb;
import teammates.storage.sqlapi.FeedbackSessionsDb;
import teammates.storage.sqlapi.NotificationsDb;
import teammates.storage.sqlapi.UsageStatisticsDb;
import teammates.storage.sqlapi.UsersDb;

/**
 * Setup in web.xml to register logic classes at application startup.
 */
public class LogicStarter implements ServletContextListener {

    private static final Logger log = Logger.getLogger();

    /**
     * Registers dependencies between different logic classes.
     */
    public static void initializeDependencies() {
        AccountsLogic accountsLogic = AccountsLogic.inst();
        AccountRequestsLogic accountRequestsLogic = AccountRequestsLogic.inst();
        CoursesLogic coursesLogic = CoursesLogic.inst();
        DataBundleLogic dataBundleLogic = DataBundleLogic.inst();
        DeadlineExtensionsLogic deadlineExtensionsLogic = DeadlineExtensionsLogic.inst();
        FeedbackSessionsLogic fsLogic = FeedbackSessionsLogic.inst();
        FeedbackResponsesLogic frLogic = FeedbackResponsesLogic.inst();
        FeedbackResponseCommentsLogic frcLogic = FeedbackResponseCommentsLogic.inst();
        FeedbackQuestionsLogic fqLogic = FeedbackQuestionsLogic.inst();
        NotificationsLogic notificationsLogic = NotificationsLogic.inst();
        UsageStatisticsLogic usageStatisticsLogic = UsageStatisticsLogic.inst();
        UsersLogic usersLogic = UsersLogic.inst();

        accountRequestsLogic.initLogicDependencies(AccountRequestsDb.inst());
        accountsLogic.initLogicDependencies(AccountsDb.inst(), notificationsLogic, usersLogic);
        coursesLogic.initLogicDependencies(CoursesDb.inst(), fsLogic, usersLogic);
        dataBundleLogic.initLogicDependencies(accountsLogic, accountRequestsLogic, coursesLogic,
                deadlineExtensionsLogic, fsLogic, fqLogic, frLogic, frcLogic,
                notificationsLogic, usersLogic);
        deadlineExtensionsLogic.initLogicDependencies(DeadlineExtensionsDb.inst(), fsLogic);
        fsLogic.initLogicDependencies(FeedbackSessionsDb.inst(), coursesLogic, frLogic, fqLogic);
        frLogic.initLogicDependencies(FeedbackResponsesDb.inst(), usersLogic, fqLogic);
        frcLogic.initLogicDependencies(FeedbackResponseCommentsDb.inst());
        fqLogic.initLogicDependencies(FeedbackQuestionsDb.inst(), coursesLogic, frLogic, usersLogic, fsLogic);
        notificationsLogic.initLogicDependencies(NotificationsDb.inst());
        usageStatisticsLogic.initLogicDependencies(UsageStatisticsDb.inst());
        usersLogic.initLogicDependencies(UsersDb.inst(), accountsLogic, frLogic, frcLogic, deadlineExtensionsLogic);
        log.info("Initialized dependencies between logic classes");
    }

    @Override
    public void contextInitialized(ServletContextEvent event) {
        // Invoked by Jetty at application startup.
        initializeDependencies();
    }

    @Override
    public void contextDestroyed(ServletContextEvent event) {
        // Nothing to do
    }

}
