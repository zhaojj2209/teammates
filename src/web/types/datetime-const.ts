export enum Hours {
    ZERO = 0,
    TWELVE = 12,
    IN_ONE_DAY = 24,
    IN_THREE_DAYS = 72,
    IN_ONE_WEEK = 168,
}

export enum Seconds {
    IN_ONE_MINUTE = 60,
}

export enum Minutes {
    IN_ONE_HOUR = 60,
}

export enum Milliseconds {
    IN_ONE_SECOND = 1000,
    IN_TEN_SECONDS = 10000,
    IN_ONE_MINUTE = 60000,
    IN_TEN_MINUTES = 600000,
    IN_FIFTEEN_MINUTES = 900000,
    IN_ONE_HOUR = 3600000,
    IN_ONE_DAY = 86400000,
    IN_ONE_WEEK = 604800000,
}

export const getDefaultDateFormat = (): DateFormat => {
    return { year: 0, month: 0, day: 0 };
};

export const getLatestTimeFormat = (): TimeFormat => {
    return { hour: 23, minute: 59 };
};

export const getDefaultTimeFormat = (): TimeFormat => {
    return { hour: 0, minute: 0 };
};

export interface DateFormat {
    year: number;
    month: number;
    day: number;
}

export interface TimeFormat {
    hour: number;
    minute: number;
}
