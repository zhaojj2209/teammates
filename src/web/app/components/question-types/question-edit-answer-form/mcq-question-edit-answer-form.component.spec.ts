import { DebugElement } from '@angular/core';
import { ComponentFixture, fakeAsync, TestBed, tick, waitForAsync } from '@angular/core/testing';
import { FormsModule } from '@angular/forms';
import { By } from '@angular/platform-browser';
import { FeedbackMcqQuestionDetails } from 'src/web/types/api-output';
import { RichTextEditorModule } from '../../rich-text-editor/rich-text-editor.module';
import { TeammatesCommonModule } from '../../teammates-common/teammates-common.module';
import { McqQuestionEditAnswerFormComponent } from './mcq-question-edit-answer-form.component';

describe('McqQuestionEditAnswerFormComponent', () => {
  let component: McqQuestionEditAnswerFormComponent;
  let fixture: ComponentFixture<McqQuestionEditAnswerFormComponent>;

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [McqQuestionEditAnswerFormComponent],
      imports: [
        FormsModule,
        RichTextEditorModule,
        TeammatesCommonModule,
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(McqQuestionEditAnswerFormComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with default view', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when questionDropdownEnabled is updated', () => {
    component.questionDetails.questionDropdownEnabled = true;
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  describe('should display all MCQ choices', () => {
    it('question dropdown disabled', () => {
      component.questionDetails.mcqChoices = ['Option 1', 'Option 2', 'Option 3'];
      fixture.detectChanges();

      const radioGroup: DebugElement[] = fixture.debugElement.queryAll(By.css('#radioOptionSpan'));
      const mcqChoices: FeedbackMcqQuestionDetails['mcqChoices'] = component.questionDetails.mcqChoices;
      const radioSpans: HTMLSpanElement[] = [];

      radioGroup.forEach((de: DebugElement) => radioSpans.push(de.nativeElement));

      for (let i = 0; i < radioSpans.length; i += 1) {
        expect(radioSpans[i].innerHTML).toStrictEqual(mcqChoices[i]);
      }
    });

    it('question dropdown enabled', () => {
      component.questionDetails.mcqChoices = ['Option 1', 'Option 2', 'Option 3'];
      component.questionDetails.questionDropdownEnabled = true;
      fixture.detectChanges();

      const select: HTMLSelectElement = fixture.debugElement.query(By.css('#dropdown-option-select')).nativeElement;

      // Removes placeholder option
      select.options.remove(0);

      const mcqChoices: FeedbackMcqQuestionDetails['mcqChoices'] = component.questionDetails.mcqChoices;

      for (let i = 0; i < select.options.length; i += 1) {
        expect(select.options[i].text).toStrictEqual(mcqChoices[i]);
      }
    });
  });

  it('should call change function upon selecting a dropdown option', fakeAsync(() => {
    const SELECTED_SELECT_INDEX = 1;

    component.questionDetails.mcqChoices = ['Option 1', 'Option 2', 'Option 3'];
    component.questionDetails.questionDropdownEnabled = true;

    fixture.detectChanges();

    const select: HTMLSelectElement = fixture.debugElement.query(By.css('#dropdown-option-select')).nativeElement;

    jest.spyOn(component, 'updateSelectedMcqDropdownOption');

    select.value = select.options[SELECTED_SELECT_INDEX].value;
    select.dispatchEvent(new Event('change'));
    tick();

    expect(component.updateSelectedMcqDropdownOption).toHaveBeenCalledTimes(1);
  }));

  it('should update answer selection upon selecting a dropdown option', () => {
    // Works for both select and mcqChoices as select first option is disabled
    const SELECTED_INDEX = 0;

    component.questionDetails.mcqChoices = ['Option 1', 'Option 2', 'Option 3'];
    component.questionDetails.questionDropdownEnabled = true;
    fixture.detectChanges();

    const select: HTMLSelectElement = fixture.debugElement.query(By.css('#dropdown-option-select')).nativeElement;

    select.value = select.options[SELECTED_INDEX].value;
    select.dispatchEvent(new Event('change'));
    fixture.detectChanges();

    expect(component.questionDetails.mcqChoices[select.selectedIndex])
      .toEqual(component.questionDetails.mcqChoices[SELECTED_INDEX]);
  });
});
