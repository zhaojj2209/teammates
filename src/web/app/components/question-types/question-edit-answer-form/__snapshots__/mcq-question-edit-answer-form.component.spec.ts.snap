// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`McqQuestionEditAnswerFormComponent should snap when questionDropdownEnabled is updated 1`] = `
<tm-mcq-question-edit-answer-form
  cssRefresh={[Function EventEmitter_]}
  id=""
  inputTextBoxOther="undefined"
  isDisabled="false"
  isMcqOptionSelected={[Function Array]}
  questionDetails={[Function Object]}
  responseDetails={[Function Object]}
  responseDetailsChange={[Function EventEmitter_]}
  valueSelected=""
>
  <tr>
    <td>
      <div>
        <select
          class="form-control margin-bottom-8px ng-untouched ng-pristine ng-valid"
          id="dropdown-option-select"
        >
          <option
            value=""
          />
        </select>
      </div>
    </td>
  </tr>
</tm-mcq-question-edit-answer-form>
`;

exports[`McqQuestionEditAnswerFormComponent should snap with default view 1`] = `
<tm-mcq-question-edit-answer-form
  cssRefresh={[Function EventEmitter_]}
  id=""
  inputTextBoxOther="undefined"
  isDisabled="false"
  isMcqOptionSelected={[Function Array]}
  questionDetails={[Function Object]}
  responseDetails={[Function Object]}
  responseDetailsChange={[Function EventEmitter_]}
  valueSelected=""
>
  <tr>
    <td>
      <div>
      </div>
    </td>
  </tr>
</tm-mcq-question-edit-answer-form>
`;
